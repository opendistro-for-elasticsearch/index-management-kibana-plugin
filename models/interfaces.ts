/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// TODO: Backend has PR out to change this model, this needs to be updated once that goes through

export interface ManagedIndexMetaData {
  index: string;
  indexUuid: string;
  policyId: string;
  policySeqNo?: number;
  policyPrimaryTerm?: number;
  policyCompleted?: boolean;
  rolledOver?: boolean;
  transitionTo?: string;
  state?: { name: string; startTime: number };
  action?: { name: string; startTime: number; index: number; failed: boolean; consumedRetries: number };
  retryInfo?: { failed: boolean; consumedRetries: number };
  info?: object;
}

/**
 * ManagedIndex item shown in the Managed Indices table
 */
export interface ManagedIndexItem {
  index: string;
  indexUuid: string;
  policyId: string;
  policySeqNo: number;
  policyPrimaryTerm: number;
  policy: Policy | null;
  enabled: boolean;
  managedIndexMetaData: ManagedIndexMetaData | null;
}

export interface IndexItem {
  index: string;
}

/**
 * Interface what the Policy Elasticsearch Document
 */
export interface DocumentPolicy {
  id: string;
  primaryTerm: number;
  seqNo: number;
  policy: Policy;
}

//Data model that contains both rollup item and metadata of rollup job
export interface DocumentRollup {
  _id: string;
  _seqNo: number;
  _primaryTerm: number;
  rollup: Rollup;
  metadata: any;
  //   {
  //   rollup_id: string;
  //   enabled: boolean;
  //   schedule: {
  //     interval?: {
  //       start_time?: number;
  //       period: number;
  //       unit: string;
  //     };
  //     cron?: {
  //       expression: string;
  //       timezone: string;
  //     };
  //   };
  //   last_updated_time: number;
  //   enabled_time: number | null;
  //   description: string;
  //   schema_version: number;
  //   source_index: string;
  //   target_index: string;
  //   metadata_id: number | null;
  //   roles: string[];
  //   page_size: number;
  //   delay: number | null;
  //   continuous: boolean;
  //   dimensions: RollupDimensionItem[];
  //   metrics: RollupMetricItem[];
  // };
}

// TODO: Fill out when needed
// TODO: separate a frontend Policy from backendPolicy
export interface Policy {
  description: string;
  default_state: string;
  states: State[];
}

export interface State {
  name: string;
  actions: object[];
  transitions: object[];
}

export interface Rollup {
  continuous: boolean;
  delay: number | null;
  description: string;
  dimensions: (DateHistogramItem | TermsItem | HistogramItem)[];
  enabled: boolean;
  enabledTime: number | null;
  lastUpdatedTime: number;
  metadata_id: string | null;
  metrics: MetricItem[];
  page_size: number;
  schedule: IntervalSchedule | CronSchedule;
  schemaVersion: number;
  source_index: string;
  target_index: string;
}

export interface RollupMetadata {
  metadata_id: string;
  rollup_metadata: {
    id: string;
    seq_no: number;
    primary_term: number;
    rollup_id: string;
    after_key: Map<String, any> | null;
    last_updated_time: number;
    continuous: {
      next_window_start_time: number;
      next_window_end_time: number;
    } | null;
    status: string;
    failure_reason: string | null;
    stats: {
      pages_processed: number | null;
      documents_processed: number | null;
      rollups_indexed: number | null;
      index_time_in_millis: number | null;
      search_time_in_millis: number | null;
    };
  };
}

export interface IntervalSchedule {
  startTime: number | null;
  period: number;
  unit: string;
}

export interface CronSchedule {
  expression: string;
  timezone: string;
}

//Frontend dimension data model
export interface DimensionItem {
  sequence: number;
  field: FieldItem;
  aggregationMethod: string;
  interval?: number;
}

//Frontend metric data model
export interface MetricItem {
  source_field: FieldItem;
  all: boolean;
  min: boolean;
  max: boolean;
  sum: boolean;
  avg: boolean;
  value_count: boolean;
}

export interface FieldItem {
  label: string;
  type?: string;
}

interface DateHistogramItem {
  sourceField: string;
  fixed_interval: string | null;
  calendar_interval: string | null;
  timezone: string;
}

interface TermsItem {
  terms: {
    source_field: string;
    target_field: string;
  };
}

interface HistogramItem {
  histogram: {
    source_field: string;
    target_field: string;
    interval: number;
  };
}

//Backend dimension data model
export type RollupDimensionItem = DateHistogramItem | TermsItem | HistogramItem;

//Backend metric data model
export interface RollupMetricItem {
  source_field: string;
  metrics: [
    {
      min?: Object;
      max?: Object;
      sum?: Object;
      avg?: Object;
      value_count?: Object;
    }
  ];
}
