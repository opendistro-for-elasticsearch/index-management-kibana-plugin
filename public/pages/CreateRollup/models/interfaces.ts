/*
 * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { ManagedCatIndex } from "../../../../server/models/interfaces";
import { FieldItem } from "../../../../models/interfaces";

export interface MappingsItem {
  properties?: {};
  type?: string;
}

export interface DimensionItem {
  sequence: number;
  field: FieldItem;
  aggregationMethod: string;
  interval?: number;
}

export interface MetricItem {
  source_field: FieldItem;
  all: boolean;
  min: boolean;
  max: boolean;
  sum: boolean;
  avg: boolean;
  value_count: boolean;
}

//From backend
export interface RollupDimensionItem {
  date_histogram?: {
    fixed_interval?: string;
    calendar_interval?: string;
    source_field: string;
    target_field: string;
    timezone: string;
  };
  terms?: {
    source_field: string;
    target_field: string;
  };
  histogram?: {
    source_field: string;
    target_field: string;
    interval: number;
  };
}

export interface RollupMetricItem {
  source_field: string;
  metrics: [
    {
      min?: Object;
      max?: Object;
      sum?: Object;
      avg?: Object;
      value_count?: Object;
    }
  ];
}
