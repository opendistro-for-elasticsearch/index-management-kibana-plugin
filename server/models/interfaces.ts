/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import { IndexService, ManagedIndexService, PolicyService, RollupService } from "../services";
import { DocumentPolicy, DocumentRollup, ManagedIndexItem, Rollup } from "../../models/interfaces";

export interface NodeServices {
  indexService: IndexService;
  managedIndexService: ManagedIndexService;
  policyService: PolicyService;
  rollupService: RollupService;
}

export interface SearchResponse<T> {
  hits: {
    total: { value: number };
    hits: { _source: T; _id: string; _seq_no?: number; _primary_term?: number }[];
  };
}

export interface ExplainResponse {
  [index: string]: ExplainAPIManagedIndexMetaData | undefined;
}

export interface GetManagedIndicesResponse {
  totalManagedIndices: number;
  managedIndices: ManagedIndexItem[];
}

export interface GetPoliciesResponse {
  policies: DocumentPolicy[];
  totalPolicies: number;
}

export interface DeletePolicyResponse {
  result: string;
}

export interface PutPolicyResponse {
  _id: string;
  // TODO: remove _version from IndexPolicyAPI
  _version: number;
  _primary_term: number;
  _seq_no: number;
  policy: { policy: object };
}

export interface DeleteRollupResponse {
  result: string;
}

export interface GetIndicesResponse {
  indices: ManagedCatIndex[];
  totalIndices: number;
}

export interface GetFieldsResponse {
  result: string;
}

export interface GetRollupsResponse {
  rollups: DocumentRollup[];
  totalRollups: number;
  metadata: any;
}

export interface PutRollupResponse {
  _id: string;
  _primary_term: number;
  _seq_no: number;
  rollup: { rollup: Rollup };
}

export interface IndexUpdateResponse {
  updatedIndices: number;
  failures: boolean;
  failedIndices: FailedIndex[];
}

export interface ApplyPolicyResponse extends IndexUpdateResponse {}

export interface RemovePolicyResponse extends IndexUpdateResponse {}

export interface ChangePolicyResponse extends IndexUpdateResponse {}

export interface RetryManagedIndexResponse extends IndexUpdateResponse {}

export interface RetryParams {
  index: string;
  body?: { state: string };
}

export interface DeletePolicyParams {
  policyId: string;
}

export interface PutPolicyParams {
  policyId: string;
  ifSeqNo?: string;
  ifPrimaryTerm?: string;
  body: string;
}
export interface DeleteRollupParams {
  rollupId: string;
}

export interface PutRollupParams {
  rollupId: string;
  if_seq_no?: string;
  if_primary_term?: string;
  body: string;
}

export interface DeleteRollupParams {
  rollupId: string;
}

export interface PutRollupParams {
  rollupId: string;
  if_seq_no?: string;
  if_primary_term?: string;
  body: string;
}

// TODO: remove optional failedIndices after fixing retry API to always array
export interface RetryResponse {
  failures: boolean;
  updated_indices: number;
  failed_indices?: BackendFailedIndex[];
}

export interface AddResponse {
  failures: boolean;
  updated_indices: number;
  failed_indices: BackendFailedIndex[];
}

export interface RemoveResponse {
  failures: boolean;
  updated_indices: number;
  failed_indices: BackendFailedIndex[];
}

export interface AcknowledgedResponse {
  acknowledged: boolean;
}

export interface BackendFailedIndex {
  index_name: string;
  index_uuid: string;
  reason: string;
}

export interface FailedIndex {
  indexName: string;
  indexUuid: string;
  reason: string;
}

export interface ExplainAPIManagedIndexMetaData {
  "opendistro.index_state_management.policy_id": string | null;
  index?: string;
  index_uuid?: string;
  policy_id?: string;
  policy_seq_no?: number;
  policy_primary_term?: number;
  policy_completed?: boolean;
  rolled_over?: boolean;
  transition_to?: string;
  state?: { name: string; start_time: number };
  action?: { name: string; start_time: number; index: number; failed: boolean; consumed_retries: number };
  retry_info?: { failed: boolean; consumed_retries: number };
  info?: object;
}

export interface IndexManagementApi {
  [API_ROUTE: string]: string;

  readonly POLICY_BASE: string;
  readonly EXPLAIN_BASE: string;
  readonly RETRY_BASE: string;
  readonly ADD_POLICY_BASE: string;
  readonly REMOVE_POLICY_BASE: string;
  readonly CHANGE_POLICY_BASE: string;
  readonly ROLLUP_JOBS_BASE: string;
}

export interface DefaultHeaders {
  "Content-Type": "application/json";
  Accept: "application/json";
}

export interface QueryStringQuery<T extends string> {
  query_string: {
    default_field: T;
    default_operator: "AND";
    query: string;
  };
}

// Default _cat index response
export interface CatIndex {
  "docs.count": string;
  "docs.deleted": string;
  health: string;
  index: string;
  pri: string;
  "pri.store.size": string;
  rep: string;
  status: string;
  "store.size": string;
  uuid: string;
}

export interface ManagedCatIndex extends CatIndex {
  managed: string;
}
