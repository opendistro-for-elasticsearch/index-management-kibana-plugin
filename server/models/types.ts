/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

export type MatchAllQuery = { match_all: {} };

export type ManagedIndicesSort = {
  [sortField: string]: string;
  index: "managed_index.index";
  policyId: "managed_index.policy_id";
};

export type PoliciesSort = {
  [sortField: string]: string;
  id: "policy.policy_id.keyword";
  "policy.policy.description": "policy.description.keyword";
  "policy.policy.last_updated_time": "policy.last_updated_time";
};

export type RollupsSort = {
  [sortField: string]: string;
  id: "rollup.rollup_id.keyword";
  "rollup.rollup.description": "rollup.description.keyword";
  "rollup.rollup.last_updated_time": "rollup.last_updated_time";
};

export type ServerResponse<T> = { ok: false; error: string } | { ok: true; response: T };
