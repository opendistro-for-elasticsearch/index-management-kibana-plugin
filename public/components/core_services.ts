import { createContext } from "react";
import { CoreStart } from "kibana/public";

const CoreServicesContext = createContext<CoreStart | null>(null);

const CoreServicesConsumer = CoreServicesContext.Consumer;

export { CoreServicesContext, CoreServicesConsumer };
