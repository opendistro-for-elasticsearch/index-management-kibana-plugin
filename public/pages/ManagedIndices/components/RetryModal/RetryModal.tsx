/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React, { Component } from "react";
import {
  EuiButton,
  EuiButtonEmpty,
  EuiModal,
  EuiModalBody,
  EuiModalFooter,
  EuiModalHeader,
  EuiModalHeaderTitle,
  EuiOverlayMask,
  EuiRadioGroup,
  EuiSelect,
  EuiSpacer,
  EuiFormRow,
} from "@elastic/eui";
import { ManagedIndexItem, State } from "../../../../../models/interfaces";
import { BrowserServices } from "../../../../models/interfaces";
import { getErrorMessage } from "../../../../utils/helpers";
import { CoreStart } from "kibana/public";

interface RetryModalProps {
  services: BrowserServices;
  retryItems: ManagedIndexItem[];
  onClose: () => void;
  core: CoreStart;
}

interface RetryModalState {
  radioIdSelected: string;
  stateSelected: string;
  stateOptions: { value: string; text: string }[];
}

enum Radio {
  Current = "current",
  State = "state",
}

export default class RetryModal extends Component<RetryModalProps, RetryModalState> {
  state = {
    radioIdSelected: Radio.Current,
    stateSelected: "",
    stateOptions: [],
  };

  componentDidMount(): void {
    this.initOptions();
  }

  initOptions = (): void => {
    const { retryItems } = this.props;
    if (!retryItems.length) return;

    // we will use this first item as the base set of states
    // if we ever encounter an item with no policy, then simply return early
    // otherwise take the intersections of all next retry item states
    // until we have a final list of common states

    const firstRetryItem = retryItems[0];
    if (!firstRetryItem.policy || !firstRetryItem.policy.states) return;
    let states = new Set(firstRetryItem.policy.states.map((state: State) => state.name));

    for (let i = 1; i < retryItems.length; i++) {
      // if we ever end up with no states just return early
      if (!states.size) return;

      const retryItem = retryItems[i];
      if (!retryItem.policy || !retryItem.policy.states) return;
      const tempStates = new Set(retryItem.policy.states.map((state: State) => state.name));

      // take intersection of two state sets
      states = new Set([...states].filter((state) => tempStates.has(state)));
    }

    const stateOptions = [...states].map((state: string) => ({ value: state, text: state }));
    this.setState({ stateOptions });
  };

  onRetry = async (): Promise<void> => {
    const { radioIdSelected, stateSelected } = this.state;
    const {
      retryItems,
      onClose,
      services: { managedIndexService },
    } = this.props;
    try {
      const indices = retryItems.map((item) => item.index);
      const state = radioIdSelected == Radio.State ? stateSelected : null;
      const response = await managedIndexService.retryManagedIndexPolicy(indices, state);
      if (response.ok) {
        const {
          response: { updatedIndices, failedIndices, failures },
        } = response;
        if (failures) {
          this.props.core.notifications.toasts.addDanger(
            `Failed to retry: ${failedIndices.map((failedIndex) => `[${failedIndex.indexName}, ${failedIndex.reason}]`).join(", ")}`
          );
        }

        if (updatedIndices) {
          this.props.core.notifications.toasts.addSuccess(`Retried ${updatedIndices} managed indices`);
        }
      } else {
        this.props.core.notifications.toasts.addDanger(response.error);
      }
      onClose();
    } catch (err) {
      this.props.core.notifications.toasts.addDanger(getErrorMessage(err, "There was a problem retrying managed indices"));
    }
  };

  onChange = (optionId: string): void => {
    this.setState({ radioIdSelected: optionId });
  };

  onSelectChange = (e: React.ChangeEvent<HTMLSelectElement>): void => {
    this.setState({ stateSelected: e.target.value });
  };

  render() {
    const { radioIdSelected, stateSelected, stateOptions } = this.state;
    const { onClose, retryItems } = this.props;

    const currentRadio = { id: Radio.Current, label: "Retry policy from current action" };
    const stateRadio = { id: Radio.State, label: "Retry policy from selected state", disabled: !stateOptions.length };
    const radioOptions = [currentRadio, stateRadio];
    return (
      <EuiOverlayMask>
        {/*
            // @ts-ignore */}
        <EuiModal onCancel={onClose} onClose={onClose}>
          <EuiModalHeader>
            <EuiModalHeaderTitle>Retry policy</EuiModalHeaderTitle>
          </EuiModalHeader>

          <EuiModalBody>
            <EuiRadioGroup options={radioOptions} idSelected={radioIdSelected} onChange={this.onChange} />

            <EuiSpacer size="s" />
            <EuiFormRow label="Start state" helpText="Only common states shared across all selected indices are available">
              <EuiSelect
                disabled={radioIdSelected !== Radio.State}
                options={stateOptions}
                value={stateSelected}
                onChange={this.onSelectChange}
                aria-label="Retry failed policy from"
              />
            </EuiFormRow>
          </EuiModalBody>

          <EuiModalFooter>
            <EuiButtonEmpty onClick={onClose} data-test-subj="retryModalCloseButton">
              Close
            </EuiButtonEmpty>

            <EuiButton onClick={this.onRetry} disabled={!retryItems.length} fill data-test-subj="retryModalRetryButton">
              Retry
            </EuiButton>
          </EuiModalFooter>
        </EuiModal>
      </EuiOverlayMask>
    );
  }
}
