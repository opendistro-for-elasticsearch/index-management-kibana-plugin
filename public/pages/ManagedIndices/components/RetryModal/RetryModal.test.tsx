/*
 * Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * or in the "license" file accompanying this file. This file is distributed
 * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 * express or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

import React from "react";
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, wait } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import RetryModal from "./RetryModal";
import { browserServicesMock, coreServicesMock } from "../../../../../test/mocks";
import { ManagedIndexItem } from "../../../../../models/interfaces";
import { CoreServicesContext } from "../../../../components/core_services";

const retryItems: ManagedIndexItem[] = [
  {
    index: "some_index",
    indexUuid: "some-index-uuid",
    policyId: "some_policy",
    policySeqNo: 1,
    policyPrimaryTerm: 1,
    policy: {
      description: "some description",
      default_state: "one",
      states: [
        { name: "one", actions: [{ delete: {} }], transitions: [] },
        { name: "two", actions: [{ delete: {} }], transitions: [] },
      ],
    },
    enabled: false,
    managedIndexMetaData: null,
  },
];

describe("<RetryModal /> spec", () => {
  it("renders the component", () => {
    render(<RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />);
    // EuiOverlayMask appends an element to the body so we should have two, an empty div from react-test-library
    // and our EuiOverlayMask element
    expect(document.body.children).toHaveLength(2);
    expect(document.body.children[1]).toMatchSnapshot();
  });

  it("calls close when close button clicked", () => {
    const onClose = jest.fn();
    const { getByTestId } = render(<RetryModal services={browserServicesMock} retryItems={retryItems} onClose={onClose} />);

    fireEvent.click(getByTestId("retryModalCloseButton"));
    expect(onClose).toHaveBeenCalled();
  });

  it("disables select from state when no common states", () => {
    const retryItems: ManagedIndexItem[] = [
      {
        index: "some_index_1",
        indexUuid: "some-index-uuid-1",
        policyId: "some_policy_1",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: {
          description: "some description",
          default_state: "one",
          states: [
            { name: "one", actions: [{ delete: {} }], transitions: [] },
            { name: "two", actions: [{ delete: {} }], transitions: [] },
            { name: "three", actions: [{ delete: {} }], transitions: [] },
          ],
        },
        enabled: false,
        managedIndexMetaData: null,
      },
      {
        index: "some_index_2",
        indexUuid: "some-index-uuid-2",
        policyId: "some_policy_2",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: {
          description: "some description",
          default_state: "five",
          states: [{ name: "five", actions: [{ delete: {} }], transitions: [] }],
        },
        enabled: false,
        managedIndexMetaData: null,
      },
      {
        index: "some_index_3",
        indexUuid: "some-index-uuid-3",
        policyId: "some_policy_3",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: {
          description: "some description",
          default_state: "two",
          states: [
            { name: "two", actions: [{ delete: {} }], transitions: [] },
            { name: "three", actions: [{ delete: {} }], transitions: [] },
          ],
        },
        enabled: false,
        managedIndexMetaData: null,
      },
    ];
    const { getByLabelText } = render(
      <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} core={coreServicesMock} />
    );

    expect(getByLabelText("Retry failed policy from")).toBeDisabled();
  });

  it("disables select from state when item has no policy", () => {
    const retryItems: ManagedIndexItem[] = [
      {
        index: "some_index_1",
        indexUuid: "some-index-uuid-1",
        policyId: "some_policy_1",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: {
          description: "some description",
          default_state: "one",
          states: [
            { name: "one", actions: [{ delete: {} }], transitions: [] },
            { name: "two", actions: [{ delete: {} }], transitions: [] },
            { name: "three", actions: [{ delete: {} }], transitions: [] },
          ],
        },
        enabled: false,
        managedIndexMetaData: null,
      },
      {
        index: "some_index_2",
        indexUuid: "some-index-uuid-2",
        policyId: "some_policy_2",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: {
          description: "some description",
          default_state: "two",
          states: [
            { name: "two", actions: [{ delete: {} }], transitions: [] },
            { name: "three", actions: [{ delete: {} }], transitions: [] },
          ],
        },
        enabled: false,
        managedIndexMetaData: null,
      },
      {
        index: "some_index_3",
        indexUuid: "some-index-uuid-3",
        policyId: "some_policy_3",
        policySeqNo: 1,
        policyPrimaryTerm: 1,
        policy: null,
        enabled: false,
        managedIndexMetaData: null,
      },
    ];
    const { getByLabelText } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />
      </CoreServicesContext.Provider>
    );

    expect(getByLabelText("Retry failed policy from")).toBeDisabled();
  });

  it("can select a different state to retry from", async () => {
    browserServicesMock.managedIndexService.retryManagedIndexPolicy = jest
      .fn()
      .mockResolvedValue({ ok: true, response: { updatedIndices: 1, failedIndices: [], failures: false } });
    const { getByLabelText, getByTestId, getByText } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        {" "}
        <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />
      </CoreServicesContext.Provider>
    );

    fireEvent.click(getByLabelText("Retry policy from selected state"));
    // @ts-ignore
    expect(getByText("two").selected).toBe(false);
    userEvent.selectOptions(getByLabelText("Retry failed policy from"), ["two"]);
    // trigger change until this is merged in: https://github.com/testing-library/user-event/pull/131
    fireEvent.change(getByLabelText("Retry failed policy from"));
    // @ts-ignore
    expect(getByText("two").selected).toBe(true);

    fireEvent.click(getByTestId("retryModalRetryButton"));

    await wait();

    expect(browserServicesMock.managedIndexService.retryManagedIndexPolicy).toHaveBeenCalledWith(["some_index"], "two");
    expect(coreServicesMock.notifications.toasts.addSuccess).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addSuccess).toHaveBeenCalledWith("Retried 1 managed indices");
  });

  it("shows error toaster when error is thrown", async () => {
    browserServicesMock.managedIndexService.retryManagedIndexPolicy = jest.fn().mockRejectedValue(new Error("this is an error"));
    const { getByTestId } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />
      </CoreServicesContext.Provider>
    );

    fireEvent.click(getByTestId("retryModalRetryButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("this is an error");
  });

  it("shows error toaster when error is returned", async () => {
    browserServicesMock.managedIndexService.retryManagedIndexPolicy = jest.fn().mockResolvedValue({ ok: false, error: "some error" });
    const { getByTestId } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />
      </CoreServicesContext.Provider>
    );

    fireEvent.click(getByTestId("retryModalRetryButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("some error");
  });

  it("shows error toaster with each failed reason", async () => {
    browserServicesMock.managedIndexService.retryManagedIndexPolicy = jest.fn().mockResolvedValue({
      ok: true,
      response: { updatedIndices: 0, failures: true, failedIndices: [{ indexName: "index_a", reason: "some reason" }] },
    });
    const { getByTestId } = render(
      <CoreServicesContext.Provider value={coreServicesMock}>
        <RetryModal services={browserServicesMock} retryItems={retryItems} onClose={() => {}} />
      </CoreServicesContext.Provider>
    );

    fireEvent.click(getByTestId("retryModalRetryButton"));

    await wait();

    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledTimes(1);
    expect(coreServicesMock.notifications.toasts.addDanger).toHaveBeenCalledWith("Failed to retry: [index_a, some reason]");
  });
});
