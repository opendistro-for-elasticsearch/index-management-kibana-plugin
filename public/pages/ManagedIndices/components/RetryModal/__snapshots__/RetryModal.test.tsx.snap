// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RetryModal /> spec renders the component 1`] = `
<div
  class="euiOverlayMask"
>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="1"
  />
  <div
    data-focus-lock-disabled="false"
  >
    <div
      class="euiModal euiModal--maxWidth-default"
      tabindex="0"
    >
      <button
        aria-label="Closes this modal window"
        class="euiButtonIcon euiButtonIcon--text euiModal__closeIcon"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="euiIcon euiIcon--medium euiButtonIcon__icon"
          focusable="false"
          height="16"
          viewBox="0 0 16 16"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7.293 8L3.146 3.854a.5.5 0 1 1 .708-.708L8 7.293l4.146-4.147a.5.5 0 0 1 .708.708L8.707 8l4.147 4.146a.5.5 0 0 1-.708.708L8 8.707l-4.146 4.147a.5.5 0 0 1-.708-.708L7.293 8z"
          />
        </svg>
      </button>
      <div
        class="euiModal__flex"
      >
        <div
          class="euiModalHeader"
        >
          <div
            class="euiModalHeader__title"
          >
            Retry policy
          </div>
        </div>
        <div
          class="euiModalBody"
        >
          <div
            class="euiModalBody__overflow"
          >
            <div>
              <div
                class="euiRadio euiRadioGroup__item"
              >
                <input
                  checked=""
                  class="euiRadio__input"
                  id="current"
                  type="radio"
                  value=""
                />
                <div
                  class="euiRadio__circle"
                />
                <label
                  class="euiRadio__label"
                  for="current"
                >
                  Retry policy from current action
                </label>
              </div>
              <div
                class="euiRadio euiRadioGroup__item"
              >
                <input
                  class="euiRadio__input"
                  id="state"
                  type="radio"
                  value=""
                />
                <div
                  class="euiRadio__circle"
                />
                <label
                  class="euiRadio__label"
                  for="state"
                >
                  Retry policy from selected state
                </label>
              </div>
            </div>
            <div
              class="euiSpacer euiSpacer--s"
            />
            <div
              class="euiFormRow"
              id="some_make_id-row"
            >
              <div>
                <label
                  class="euiFormLabel"
                  for="some_make_id"
                >
                  Start state
                </label>
              </div>
              <div
                class="euiFormControlLayout"
              >
                <div
                  class="euiFormControlLayout__childrenWrapper"
                >
                  <select
                    aria-describedby="some_make_id-help"
                    aria-label="Retry failed policy from"
                    class="euiSelect"
                    disabled=""
                    id="some_make_id"
                  >
                    <option
                      value="one"
                    >
                      one
                    </option>
                    <option
                      value="two"
                    >
                      two
                    </option>
                  </select>
                  <div
                    class="euiFormControlLayoutIcons euiFormControlLayoutIcons--right"
                  >
                    <span
                      class="euiFormControlLayoutCustomIcon"
                    >
                      <svg
                        aria-hidden="true"
                        class="euiIcon euiIcon--medium euiFormControlLayoutCustomIcon__icon"
                        focusable="false"
                        height="16"
                        viewBox="0 0 16 16"
                        width="16"
                        xmlns="http://www.w3.org/2000/svg"
                        xmlns:xlink="http://www.w3.org/1999/xlink"
                      >
                        <defs>
                          <path
                            d="M13.069 5.157L8.384 9.768a.546.546 0 0 1-.768 0L2.93 5.158a.552.552 0 0 0-.771 0 .53.53 0 0 0 0 .759l4.684 4.61c.641.631 1.672.63 2.312 0l4.684-4.61a.53.53 0 0 0 0-.76.552.552 0 0 0-.771 0z"
                            id="arrow_down-a"
                          />
                        </defs>
                        <use
                          fill-rule="nonzero"
                          xlink:href="#arrow_down-a"
                        />
                      </svg>
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="euiFormHelpText euiFormRow__text"
                id="some_make_id-help"
              >
                Only common states shared across all selected indices are available
              </div>
            </div>
          </div>
        </div>
        <div
          class="euiModalFooter"
        >
          <button
            class="euiButtonEmpty euiButtonEmpty--primary"
            data-test-subj="retryModalCloseButton"
            type="button"
          >
            <span
              class="euiButtonEmpty__content"
            >
              <span
                class="euiButtonEmpty__text"
              >
                Close
              </span>
            </span>
          </button>
          <button
            class="euiButton euiButton--primary euiButton--fill"
            data-test-subj="retryModalRetryButton"
            type="button"
          >
            <span
              class="euiButton__content"
            >
              <span
                class="euiButton__text"
              >
                Retry
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <div
    data-focus-guard="true"
    style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
    tabindex="0"
  />
</div>
`;
